golang/pkg/rnr/tasks.go

```diff
@@ -28,10 +33,84 @@ func taskSchedState(pbt *pb.Task) TaskState {
 	return UNKNOWN
 }
 
+// TaskCallback is installed by higher-level construct, such as callback task or nested task.
+type TaskCallback func(context.Context, *Task)
+type StateUpdateCallback func(*pb.Task) *pb.Task
+
 // Task is a generic interface for pollable tasks
-type Task interface {
-	Poll()
-	Proto(updater func(*pb.Task)) *pb.Task
-	SetState(pb.TaskState)
-	GetChild(name string) Task
+type Task struct {
+	cb           TaskCallback
+	pb           *pb.Task
+	children     []*Task
+	has_children bool
```

Comment:
You don't seem to be using this field anywhere, so is there really a need for this? Also this could be easily computed as `has_children := len(t.children) > 0`.

It would also make the `NewTask` constructor below simpler, as you don't need to pass a boolean.

Reply:
This is actually supposed to work the other way around -- since we turned the `Task` into a struct, any manipulation w.r.t. children needs to be implemented only there. `has_children` is a flag that indicates whether `Add()` should be able to add new children. The check for this was missing, but I'm adding it now.

Having has_children as a constructor parameter makes it slightly more verbose than it should be. Do you think this would warrant a separate constructor? (i.e. NewTask and NewTaskWithChildren)